xbox/xbox.c

```diff
@@ -0,0 +1,1002 @@
+           operand == GL_ONE_MINUS_SRC_ALPHA);
+    invert ^= (operand == GL_ONE_MINUS_SRC_ALPHA);
+  }
+
+  return invert ? _RC_UNSIGNED_INVERT : _RC_UNSIGNED;
+}
+
+static uint32_t* setup_texenv_src(uint32_t* p,
+                                  Texture* t, unsigned int texture, bool rgb,
+                                  int a_arg, bool a_invert,
+                                  int b_arg, bool b_invert,
+                                  int c_arg, bool c_invert,
+                                  int d_arg, bool d_invert) {
+
+  //FIXME: Set var to whatever the arg demands
+
+
+  unsigned int x_a = gl_to_texenv_src(t, texture, rgb, a_arg);
+  unsigned int x_b = gl_to_texenv_src(t, texture, rgb, b_arg);
+  unsigned int x_c = gl_to_texenv_src(t, texture, rgb, c_arg);
+  unsigned int x_d = gl_to_texenv_src(t, texture, rgb, d_arg);
+
+  unsigned int x_a_alpha = gl_to_texenv_src_alpha(t, texture, rgb, a_arg);
+  unsigned int x_b_alpha = gl_to_texenv_src_alpha(t, texture, rgb, b_arg);
+  unsigned int x_c_alpha = gl_to_texenv_src_alpha(t, texture, rgb, c_arg);
+  unsigned int x_d_alpha = gl_to_texenv_src_alpha(t, texture, rgb, d_arg);
+
+  unsigned int x_a_map = gl_to_texenv_src_map(t, texture, rgb, a_arg, a_invert);
+  unsigned int x_b_map = gl_to_texenv_src_map(t, texture, rgb, b_arg, b_invert);
+  unsigned int x_c_map = gl_to_texenv_src_map(t, texture, rgb, c_arg, c_invert);
+  unsigned int x_d_map = gl_to_texenv_src_map(t, texture, rgb, d_arg, d_invert);
+
+  if (rgb) {
+    p = pb_push1(p, NV097_SET_COMBINER_COLOR_ICW + texture * 4,
+          MASK(NV097_SET_COMBINER_COLOR_ICW_A_SOURCE, x_a) | MASK(NV097_SET_COMBINER_COLOR_ICW_A_ALPHA, x_a_alpha) | MASK(NV097_SET_COMBINER_COLOR_ICW_A_MAP, x_a_map)
+        | MASK(NV097_SET_COMBINER_COLOR_ICW_B_SOURCE, x_b) | MASK(NV097_SET_COMBINER_COLOR_ICW_B_ALPHA, x_b_alpha) | MASK(NV097_SET_COMBINER_COLOR_ICW_B_MAP, x_b_map)
+        | MASK(NV097_SET_COMBINER_COLOR_ICW_C_SOURCE, x_c) | MASK(NV097_SET_COMBINER_COLOR_ICW_C_ALPHA, x_c_alpha) | MASK(NV097_SET_COMBINER_COLOR_ICW_C_MAP, x_c_map)
+        | MASK(NV097_SET_COMBINER_COLOR_ICW_D_SOURCE, x_d) | MASK(NV097_SET_COMBINER_COLOR_ICW_D_ALPHA, x_d_alpha) | MASK(NV097_SET_COMBINER_COLOR_ICW_D_MAP, x_d_map));
+  } else {
+    p = pb_push1(p, NV097_SET_COMBINER_ALPHA_ICW + texture * 4,
+          MASK(NV097_SET_COMBINER_ALPHA_ICW_A_SOURCE, x_a) | MASK(NV097_SET_COMBINER_ALPHA_ICW_A_ALPHA, x_a_alpha) | MASK(NV097_SET_COMBINER_ALPHA_ICW_A_MAP, x_a_map)
+        | MASK(NV097_SET_COMBINER_ALPHA_ICW_B_SOURCE, x_b) | MASK(NV097_SET_COMBINER_ALPHA_ICW_B_ALPHA, x_b_alpha) | MASK(NV097_SET_COMBINER_ALPHA_ICW_B_MAP, x_b_map)
+        | MASK(NV097_SET_COMBINER_ALPHA_ICW_C_SOURCE, x_c) | MASK(NV097_SET_COMBINER_ALPHA_ICW_C_ALPHA, x_c_alpha) | MASK(NV097_SET_COMBINER_ALPHA_ICW_C_MAP, x_c_map)
+        | MASK(NV097_SET_COMBINER_ALPHA_ICW_D_SOURCE, x_d) | MASK(NV097_SET_COMBINER_ALPHA_ICW_D_ALPHA, x_d_alpha) | MASK(NV097_SET_COMBINER_ALPHA_ICW_D_MAP, x_d_map));
+  }
```

Comment:
This entire if / else garbage should probably be part of XGUX with underlying crap in XGU
CC @dracc